[WebXR] Implicitly clear WebXROpaqueFramebuffer

djg opened the pull request:
#### 729393d0741fe040579bf599843215b206d0b6b1
<pre>
[WebXR] Implicitly clear WebXROpaqueFramebuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=274136">https://bugs.webkit.org/show_bug.cgi?id=274136</a>
<a href="https://rdar.apple.com/128041214">rdar://128041214</a>

Reviewed by Kimmo Kinnunen and Mike Wyrzykowski.

The performance panel of GPU debugger shows 3.5ms spent just clearing and
writing render targets at the start of frame due to our explicit glClear to
match the WebXR spec.

This change makes use of the robust resource feature of ANGLE to initialize
render target attachments just before their first use. To achieve this,
framebuffer discard/invalidate now marks attachments as needing initialization
when robust resource init is enabled is ANGLE.

This change adds framebufferDiscard and IPC messages gated on availability of
WebXR for WebXROpaqueFramebuffer to use.

* Source/ThirdParty/ANGLE/src/libANGLE/Framebuffer.cpp:
(gl::Framebuffer::discard):
(gl::Framebuffer::invalidate):
(gl::Framebuffer::markDrawAttachmentsNeedInit):
* Source/ThirdParty/ANGLE/src/libANGLE/Framebuffer.h:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::blitSharedToLayered):
* Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
(WebCore::GraphicsContextGL::framebufferDiscard):
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::framebufferDiscard):
(WebCore::GraphicsContextGLCocoa::enableRequiredWebXRExtensionsImpl):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(framebufferDiscard):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::framebufferDiscard):

Canonical link: <a href="https://commits.webkit.org/280236@main">https://commits.webkit.org/280236@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ec9f4040fe6f37991db2784061370761d8180323

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54475 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33888 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7039 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57754 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5206 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56778 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41422 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5196 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44112 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3490 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56570 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32023 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47176 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25248 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28821 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4506 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3349 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/50632 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4719 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59345 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29701 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4877 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51536 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30854 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47266 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50906 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12418 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31835 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30646 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple reviewed (changes requested):
I don't think we can make these changes without extensive care. Nor do reviews, for that matter.

ANGLE part should be reviewed in upstream ANGLE and pulled to downstream. This is mandatory at somepoint anyway. It wastes the least amount of time overall for everybody when it is done from the get go.

The implementation should have tests.

The ANGLE part should be tested in angle/src/tests/gl_tests/WebGLCompatibilityTest.cpp

The tests could cover at least some cases like:
 - readpixels
 - detaching textures and reading via different means
 - drawing partially / with blend on top of invalidated framebuffers
 - shader framebuffer reads
 
Naturally the tests should ensure that no unexpected pixel/stencil/depth values are observed

If the patch implements DiscardFramebuffer and InvalidateFramebuffer, the difference of them should probably be visible in the patch and perhaps even tested.

For the WebKit part, if the speedup is there (the message does not say) I'm not convinced yet the implementation is correct. 

To me, it would appear that it does not matter who does the clear: the caller explicitly or the driver implementation implicitly. Logically they should take the same amount of time. The driver would be able to skip the clear if it can prove that the combined draws after it touches all the pixels. I don't believe ANGLE has any such mechanism generally. If the speedup is there, it so that we just skip the clear, violating the intended security measures?

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/libANGLE/Framebuffer.cpp:
```diff
@@ -2896,4 +2900,39 @@ angle::Result Framebuffer::syncAttachmentState(const Context *context,
+        case GL_DEPTH_EXT:
+            if (mState.mDepthAttachment.isAttached())
+            {
+                mState.mResourceNeedsInit.set(DIRTY_BIT_DEPTH_ATTACHMENT);
```
Do you need to mark the resource itself as needing init?

E.g. if the resource is read or written through this Framebuffer, the framebuffer state mResourceNeedsInit should probably take care of the initialization.

However, if the resource is bound to some other fbo and then read through it, or  read through it some other means (shader samplling), then ANGLE cannot understand to clear it?

Does this need   `attachment.setInitState(InitState::MayNeedInit);` per the attachment?

djg replied:
It also needs to notify the framebuffer observers about the dirty state.

kkinnunen-apple commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_state_cache.h:
```diff
@@ -359,7 +359,7 @@ struct RenderPassDepthAttachmentDesc : public RenderPassAttachmentDesc
     }
 
-    double clearDepth = 0;
+    double clearDepth = 1.0;
```
Does this change the default depth clear value from 0.0 to 1.0? I don't think we can do that, that would break WebGL?

djg replied:
Although I don't think it hurts since the default value of a mtl render pass depth attachment clearDepth is 1

kkinnunen-apple reviewed (commented):
Did you btw look at ANGLE_MTL_SIMULATE_DISCARD_FRAMEBUFFER? I think that's intended to put random to the draw result when discarding, probably to make it easier to see cases where the random results aren't expected but still appear?

kkinnunen-apple commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
```diff
@@ -162,27 +161,15 @@ void WebXROpaqueFramebuffer::startFrame(const PlatformXR::FrameData::LayerData&
         m_completionSyncEvent = MachSendRight(data.layerSetup->completionSyncEvent);
     }
 
+    gl->bindFramebuffer(GL::FRAMEBUFFER, m_drawFramebuffer->object());
```
The foveation setup is very buggy, if it still needs the correct framebuffer..
if you move the bind here, the setupFramebuffer might not have a expected framebuffer bound.

djg committed:
[WebXR] Implicitly clear WebXROpaqueFramebuffer
https://bugs.webkit.org/show_bug.cgi?id=274136
rdar://128041214

Reviewed by Kimmo Kinnunen and Mike Wyrzykowski.

The performance panel of GPU debugger shows 3.5ms spent just clearing and
writing render targets at the start of frame due to our explicit glClear to
match the WebXR spec.

This change makes use of the robust resource feature of ANGLE to initialize
render target attachments just before their first use. To achieve this,
framebuffer discard/invalidate now marks attachments as needing initialization
when robust resource init is enabled is ANGLE.

This change adds framebufferDiscard and IPC messages gated on availability of
WebXR for WebXROpaqueFramebuffer to use.

* Source/ThirdParty/ANGLE/src/libANGLE/Framebuffer.cpp:
(gl::Framebuffer::discard):
(gl::Framebuffer::invalidate):
(gl::Framebuffer::markDrawAttachmentsNeedInit):
* Source/ThirdParty/ANGLE/src/libANGLE/Framebuffer.h:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::blitSharedToLayered):
* Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
(WebCore::GraphicsContextGL::framebufferDiscard):
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::framebufferDiscard):
(WebCore::GraphicsContextGLCocoa::enableRequiredWebXRExtensionsImpl):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(framebufferDiscard):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::framebufferDiscard):

Canonical link: https://commits.webkit.org/280236@main
